### PR TITLE
#21 highlight changed register

### PR DESCRIPTION
**Pull request by timxie140**
Implement highlight changed register function and remove highlight at reset in RegisterTable.java, and add connection functions in simulator.java and window.java

### CODE REVIEW

File: src/main/java/com/ezasm/gui/RegisterTable.java
Comment (TrevorBrunette) [inline src/main/java/com/ezasm/gui/RegisterTable.java:57]: These should use theme colors -- not arbitrary ones.

File: src/main/java/com/ezasm/simulation/Simulator.java
Comment (TrevorBrunette) [inline src/main/java/com/ezasm/simulation/Simulator.java:119]: This should not be in the Simulator but rather the SimulationGUI class which handles GUI simulations since this is solely a GUI feature.

File: src/main/java/com/ezasm/parsing/Line.java
Comment (TrevorBrunette) [inline src/main/java/com/ezasm/parsing/Line.java:65]: We cannot guarantee that the first register found will be an output for every instruction so this does not work for those odd instructions. A different approach could be to make changes to the `set` functions in the `Register` class which trigger the highlighting.

**Review by TrevorBrunette — CHANGES_REQUESTED** [review]
See comments

File: src/main/java/com/ezasm/gui/RegisterTable.java
Comment (TrevorBrunette) [inline src/main/java/com/ezasm/gui/RegisterTable.java:48]: This javadoc comment is improperly formatted and contains characters we don't use -- `NBSP`

Comment (timxie140) [reply src/main/java/com/ezasm/gui/RegisterTable.java:48]: I believe this is changed by the maven formatter

Comment (timxie140) [reply src/main/java/com/ezasm/gui/RegisterTable.java:48]: I'll look into it

**Comment by WyattRoss** [comment]
Another couple things I noticed is that `$pc` is just always highlighted. I'm not sure that ought to be the case.

File: src/main/java/com/ezasm/gui/RegisterTable.java
Comment (timxie140) [reply src/main/java/com/ezasm/gui/RegisterTable.java:57]: Then is it ok if I use the red in theme?

File: src/main/java/com/ezasm/gui/RegisterTable.java
Comment (WyattRoss) [inline src/main/java/com/ezasm/gui/RegisterTable.java:55]: getting the theme every time seems weird. Why not store the colors we need when we call `applyTheme`?

File: src/main/java/com/ezasm/gui/RegisterTable.java
Comment (WyattRoss) [inline src/main/java/com/ezasm/gui/RegisterTable.java:161]: remove extra what? Add javadoc to describe purpose, possibly rename this method as well.

File: src/main/java/com/ezasm/gui/Window.java
Comment (WyattRoss) [inline src/main/java/com/ezasm/gui/Window.java:340]: You're resetting the config every time. That's not needed, you only every use the copy of config you save to get a theme from it. We get the theme directly passed to the table class with `applyTheme`. Either save the whole theme there or just the colors you use from it. This will still properly update with changing the theme while the program is running.

File: src/main/java/com/ezasm/parsing/Line.java
Comment (WyattRoss) [inline src/main/java/com/ezasm/parsing/Line.java:124]: This method, and the `registerNum` variable in general seem to not be used at all. Consider removing

File: src/main/java/com/ezasm/simulation/Register.java
Comment (WyattRoss) [inline src/main/java/com/ezasm/simulation/Register.java:74]: we don't pass data around as `byte[]`s anymore. There's a method call above this which uses a new class `RawData`. Add add your changes to that method as well.

**Review by WyattRoss — CHANGES_REQUESTED** [review]
see comments. Also, highlighting in general didn't work on my machine. Not sure why, and I don't have time to figure it out

**Review by WyattRoss — DISMISSED** [review]
I removed a print statement left hanging around, but this looks like it fixes all the issues I spotted, nice work. Squash and merge!

**Comment by timxie140** [comment]
Please don't merge it yet, I'm still figuring out more test cases to make sure it will definitely work after the merge!

**Comment by timxie140** [comment]
I've tested for the examples in EzASM example folders and nothing goes wrong, also the basic cases I wrote myself passed too if you have any trouble testing it on your machine please let me know, otherwise, I think it's ready for merge!

File: src/main/java/com/ezasm/gui/RegisterTable.java
Comment (TrevorBrunette) [inline src/main/java/com/ezasm/gui/RegisterTable.java:151]: Why is `reset` an `int` if it is used as a `boolean`?

File: src/main/java/com/ezasm/gui/RegisterTable.java
Comment (TrevorBrunette) [inline src/main/java/com/ezasm/gui/RegisterTable.java:61]: What is the purpose of checking the array size here?

File: src/main/java/com/ezasm/simulation/Simulator.java
Comment (TrevorBrunette) [inline src/main/java/com/ezasm/simulation/Simulator.java:74]: This code should not be in the Simulator class, it should only be in GUI elements that cause a `reset` on the simulator (e.g., pressing the `reset` button).

**Review by TrevorBrunette — CHANGES_REQUESTED** [review]
See comments

**Comment by timxie140** [comment]
I notice that I fetch from main and then every button I click in the program will result in the following error, and none of them works, think it's caused by the AMD platform 
WeChat Image_20230301095731

File: src/main/java/com/ezasm/simulation/Register.java
Comment (TrevorBrunette) [inline src/main/java/com/ezasm/simulation/Register.java:71]: We have a function in Window that checks this condition: try `if (Window.hasInstance())`

**Comment by TrevorBrunette** [comment]
> I notice that I fetch from main and then every button I click in the program will result in the following error, and none of them works, think it's caused by the AMD platform WeChat Image_20230301095731

This is not a fault of this branch, it seems to be because of [this other PR](https://github.com/TrevorBrunette/EzASM/pull/109). You do not need to worry about this issue since it is not caused by your code. We will try and get a hotfix out soon.

**Comment by timxie140** [comment]
I changed the code and it should work properly but because of the AMD issue I can't test it on my machine for now.

**Comment by timxie140** [comment]
Please let me know if I need to make further modifications. Or should I just click the merge pull request?

**Comment by timxie140** [comment]
ok, I'll take a look of the website!

**Review by TrevorBrunette — APPROVED** [review]
LGTM, time to squash and merge

**Comment by TrevorBrunette** [comment]
You just pushed onto main using a normal merge putting many unnecessary refs onto main. I am going to have to revert it so that the git log is refs aren't cluttered by this. `Squash merge` is not the same as `Merge`.

**Pull request merged**
